test/integration/options/watch.spec.js

```diff
@@ -111,50 +63,90 @@ describe('--watch', function() {
+      const testFile = path.join(this.tempDir, 'test.js');
+      copyFixture('options/watch/test-file-change', testFile);
+
+      return runMochaWatch([testFile], this.tempDir, () => {
+        replaceFileContents(
+          testFile,
+          'testShouldFail = true',
+          'testShouldFail = false'
+        );
+      }).then(results => {
+        expect(results, 'to have length', 2);
+        expect(results[0].passes, 'to have length', 0);
+        expect(results[0].failures, 'to have length', 1);
+        expect(results[1].passes, 'to have length', 1);
+        expect(results[1].failures, 'to have length', 0);
+      });
+    });
+
+    it('reloads test dependencies when they change', function() {
+      const testFile = path.join(this.tempDir, 'test.js');
+      copyFixture('options/watch/test-with-dependency', testFile);
+
+      const dependency = path.join(this.tempDir, 'lib', 'dependency.js');
+      copyFixture('options/watch/dependency', dependency);
+
+      return runMochaWatch([testFile], this.tempDir, () => {
+        replaceFileContents(
+          dependency,
+          'module.exports.testShouldFail = false',
+          'module.exports.testShouldFail = true'
+        );
+      }).then(results => {
+        expect(results, 'to have length', 2);
+        expect(results[0].passes, 'to have length', 1);
+        expect(results[0].failures, 'to have length', 0);
+        expect(results[1].passes, 'to have length', 0);
+        expect(results[1].failures, 'to have length', 1);
+      });
     });
   });
 });
 
 /**
- * Invokes the mocha binary with the `--watch` argument for the given fixture.
+ * Runs the mocha binary in watch mode calls `change` and returns the
+ * JSON reporter output.
  *
- * Returns child process and a promise for the test results. The test results
- * are an array of JSON objects generated by the JSON reporter.
+ * The function starts mocha with the given arguments and `--watch` and
+ * waits until the first test run has completed. Then it calls `change`
+ * and waits until the second test run has been completed. Mocha is
+ * killed and the list of JSON outputs is returned.
  */
-function runMochaJSONWatchAsync(args, spawnOpts) {
-  args = [...args, '--watch'];
-  const [mocha, mochaDone] = runMochaJSONRawAsync(args, spawnOpts);
-  const testResults = mochaDone.then(data => {
-    const testResults = data.output
-      // eslint-disable-next-line no-control-regex
-      .replace(/\u001b\[\?25./g, '')
-      .split('\u001b[2K')
-      .map(x => JSON.parse(x));
-    return testResults;
-  });
-  return [mocha, testResults];
+function runMochaWatch(args, cwd, change) {
+  const [mochaProcess, resultPromise] = helpers.invokeMochaAsync(
+    [...args, '--watch', '--reporter', 'json'],
+    {cwd}
+  );
+
+  return sleep(1000)
+    .then(() => change())
+    .then(() => sleep(1000))
+    .then(() => {
+      mochaProcess.kill('SIGINT');
+      return resultPromise.then(data => {
```

Comment:
this promise needn't be nested. just return `resultPromise` and continue the chain from L139